File: scripts/research/bench.py
import json, sys, time
import ahocorasick
patterns = open(sys.argv[1]).read().splitlines()
text = open(sys.argv[2]).read()
t = time.perf_counter_ns()
ac = ahocorasick.Automaton()
for i, p in enumerate(patterns): ac.add_word(p, (i, len(p)))
ac.make_automaton()
build = (time.perf_counter_ns() - t) / 1e6
def scan():
    count = checksum = 0
    for end, (i, length) in ac.iter(text):
        count += 1
        checksum += end - length + 1 + end + 1 + i
    return count, checksum
for _ in range(20):
    scan()
    next(ac.iter(text), None)
t = time.perf_counter_ns(); count, checksum = scan(); query = (time.perf_counter_ns() - t) / 1e6
t = time.perf_counter_ns(); next(ac.iter(text), None); first = (time.perf_counter_ns() - t) / 1e6
print(json.dumps(dict(buildMs=build, queryMs=query, firstMs=first, count=count, checksum=checksum, reportedAutomatonBytes=ac.get_stats()['total_size'])))

File: scripts/research/bench.rs
use std::{env, fs, hint::black_box, time::Instant};
fn main() {
    let args: Vec<String> = env::args().collect();
    let patterns: Vec<String> = fs::read_to_string(&args[2])
        .unwrap()
        .lines()
        .map(str::to_owned)
        .collect();
    let text = fs::read_to_string(&args[3]).unwrap();
    let before = Instant::now();
    if args[1] == "rust" {
        let ac = aho_corasick::AhoCorasick::new(&patterns).unwrap();
        let build = before.elapsed().as_secs_f64() * 1000.0;
        let scan = || {
            ac.find_overlapping_iter(&text)
                .fold((0u64, 0u64), |(n, s), m| {
                    (
                        n + 1,
                        s + m.start() as u64 + m.end() as u64 + m.pattern().as_usize() as u64,
                    )
                })
        };
        for _ in 0..20 {
            black_box(scan());
            black_box(ac.find_overlapping_iter(&text).next());
        }
        let t = Instant::now();
        let result = scan();
        let query = t.elapsed().as_secs_f64() * 1000.0;
        let t = Instant::now();
        black_box(ac.find_overlapping_iter(&text).next());
        let first = t.elapsed().as_secs_f64() * 1000.0;
        println!("{{\"buildMs\":{},\"queryMs\":{},\"firstMs\":{},\"count\":{},\"checksum\":{},\"reportedAutomatonBytes\":{}}}",build,query,first,result.0,result.1,ac.memory_usage());
    } else {
        let ac = daachorse::DoubleArrayAhoCorasick::<u32>::new(&patterns).unwrap();
        let build = before.elapsed().as_secs_f64() * 1000.0;
        let scan = || {
            ac.find_overlapping_iter(&text)
                .fold((0u64, 0u64), |(n, s), m| {
                    (
                        n + 1,
                        s + m.start() as u64 + m.end() as u64 + m.value() as u64,
                    )
                })
        };
        for _ in 0..20 {
            black_box(scan());
            black_box(ac.find_overlapping_iter(&text).next());
        }
        let t = Instant::now();
        let result = scan();
        let query = t.elapsed().as_secs_f64() * 1000.0;
        let t = Instant::now();
        black_box(ac.find_overlapping_iter(&text).next());
        let first = t.elapsed().as_secs_f64() * 1000.0;
        println!(
            "{{\"buildMs\":{},\"queryMs\":{},\"firstMs\":{},\"count\":{},\"checksum\":{}}}",
            build, query, first, result.0, result.1
        );
    }
}
